packages/server/src/loop.ts

```diff
@@ -34,160 +39,161 @@ export async function chargeCustomerInvoice({
-          populate: ['project', 'items', 'subscription.changes', 'subscription.customer'],
+          populate: ['project', 'changes', 'customer'],
         },
       );
 
-      for await (let invoice of invoices) {
-        // Lock invoice processing
-        invoice.status = 'pending';
-        await database.em.persistAndFlush(invoice);
+      for await (const subscription of subscriptions) {
+        // TODO: should we lock subscription processing?
 
-        const { project, subscription } = invoice;
-        if (!subscription) {
-          log.error({ invoiceId: invoice._id }, 'Invoice has no subscription');
-          throw new Error('Invoice has no subscription');
-        }
+        const { project, customer } = subscription;
 
-        const { customer } = subscription;
+        const billingPeriod = getPeriodFromAnchorDate(subscription.nextPayment, subscription.anchorDate);
 
         try {
-          await database.em.populate(customer, ['activePaymentMethod']);
-          if (!customer.activePaymentMethod) {
-            log.error({ invoiceId: invoice._id, customerId: customer._id }, 'Customer has no active payment method');
-            throw new Error('Customer has no active payment method');
+          const existingInvoices = await database.invoices.find({
+            date: { $gte: billingPeriod.start, $lte: billingPeriod.end },
+            subscription,
+          });
+          if (existingInvoices.length > 0) {
+            log.error(
+              { subscriptionId: subscription._id, customerId: customer._id },
+              'Invoice for period already exists',
+            );
+            // TODO: should we just ignore this? currently this sets the subscription to error state?
+            throw new Error('Invoice for period already exists');
           }
 
-          invoice = addSubscriptionChangesToInvoice(subscription, invoice);
-          await chargeCustomerInvoice({ customer, invoice });
+          customer.invoiceCounter += 1;
+
+          const invoice = new Invoice({
+            currency: project.currency,
+            vatRate: project.vatRate,
+            sequentialId: customer.invoiceCounter,
+            subscription,
+            project,
+            status: 'draft',
+            date: billingPeriod.end, // TODO: or has this to be the current date when the invoice is issued?
+          });
+
+          const period = new SubscriptionPeriod(subscription, billingPeriod.start, billingPeriod.end);
+          period.getInvoiceItems().forEach((item) => {
+            invoice.items.add(item);
+          });
+
+          await database.em.persistAndFlush([customer, invoice]);
+
+          await chargeCustomerInvoice({ billingPeriod, customer, invoice });
+
+          const nextPeriod = getPeriodFromAnchorDate(
+            dayjs(subscription.nextPayment).add(1, 'day').toDate(),
+            subscription.anchorDate,
+          );
+          subscription.nextPayment = nextPeriod.end;
```

Comment:
@anbraten Here I would either take `dayjs(nextPeriod.end).add(1, 'day').startOf('day')`  or we need to change the query at the top of the method from `{ nextPayment: { $lte: now } }` to `{ nextPayment: { $lt: now } }`. It is only 1 ms but it is easier to understand later on if we do it properly.

Reply:
Changed it so we charge one day after the period ended.

Reply:
@anbraten But you missed calling `startOf('day')`. With this implementation you have the period end at `31.01. 23:59:59` and the nextPayment is therefor set to `1.2. 23:59:59`  which is somehow incorrect.

And shouldn't be `dayjs(subscription.nextPayment).toDate()` enough in the line 159 ?

Reply:
It has to be like this, right: 
![image](https://github.com/geprog/gringotts/assets/6918444/b82ccc27-4c8e-4d04-b112-18cbcdcb481a)

So we always want `nextPayment` to be `period.end + 1 hour buffer` and then I also have to subtract the buffer / get the previous period when getting the billing period in the loop?

Reply:
@anbraten Yes let's do it like that.

Reply:
@mariusheine Adjusted the code a bit again. Hopefully its correct now.

Reply:
In my test with I got the following:

```
anchorDate: 2020-01-01T07:23:00.000Z 
1. period: 
      - start: 2020-01-01T00:00:00.000Z
      - end: 2020-01-31T23:59:59.999Z
1. nextPayment: 2020-02-01T00:00:00.000Z 
2. period: 
     - start: 2020-03-01T00:00:00.000Z, 
     - end: 2020-03-31T23:59:59.999Z
6. nextPayment: 2020-03-01T00:00:00.000Z
```

Reply:
@anbraten Perfect